### DISCUSSION TITLE
Prevent multiple space characters after mention node

### DISCUSSION THREAD

**Discussion by gambolputty** [ueberdosis/tiptap#1703] [Feature Requests]
Right now a space character is inserted after every mention:
https://github.com/ueberdosis/tiptap/blob/ffd241b2d898951205ecb5f5bbda474fc839840e/packages/extension-mention/src/mention.ts#L25-L38

Maybe that's because it's assumed that you only use this extension at the end of your text document (?)
But if you use this extension in the middle of your document, this can become annoying if you accidentally inserted a node before a space character (it would insert another space character and you would have to remove the second space manually afterwards).

One could check to see if there's a text node after the caret and if it's starting with a space characters. In this case, we can omit the space:

```js
{
...
  command: ({ editor, range, props }) => {
    // don't add another space characters when there's already one after the caret
    const nodeAfter = editor.view.state.selection.$to.nodeAfter
    const addSpace = !(nodeAfter !== null && nodeAfter.type.name === 'text' && nodeAfter.text.startsWith(' '))
    const content = [
      {
        type: 'mention',
        attrs: props
      }
    ]

    if (addSpace) {
      content.push({
        type: 'text',
        text: ' '
      })
    }

    editor
      .chain()
      .focus()
      .insertContentAt(range, content)
      .run()
  },
...
}
```

**Comment by gambolputty** [comment]
@philippkuehn suggested a different strategy (increasing the range by one when the next text node starts with a space character).

Does that look better?

```js
  command: ({ editor, range, props }) => {
    // increase range.to by one when the next node is of type "text"
    // and starts with a space character
    const nodeAfter = editor.view.state.selection.$to.nodeAfter
    const overrideSpace = nodeAfter?.text?.startsWith(' ')

    if (overrideSpace) {
      range.to += 1
    }

    editor
      .chain()
      .focus()
      .insertContentAt(range, [
        {
          type: 'symbol',
          attrs: props
        },
        {
          type: 'text',
          text: ' '
        }
      ])
      .run()
  },
```

Strange observation: while developing, the text node after started with a _non-breaking space_ once and `.startsWith(' ')` didn't work. I don't know why. Should we instead check with `text.charCodeAt(0) === 32 || text.charCodeAt(0) === 160` (160 being non-breaking space)? Although the NBS would be replaced by a regular space then. Can NBS characters be intentional?

**Reply by philippkuehn** [reply]
yeah that’s better! I’m checking for any type of whitespace for now.

```js
const overrideSpace = nodeAfter?.text
  ? /^\s/.test(nodeAfter.text)
  : false
```

**Reply by gambolputty** [reply]
Isn't that a problem, because '\s' also matches "tab, form feed, line feed, and other Unicode spaces"? [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Character_Classes)

All these other space characters (e.g. line break?) would be replaced by a regular space then (?)

**Reply by gambolputty** [reply]
Oh, wait. Line breaks are handled by the HardBreak extension! Don't know about the other characters though...

**Reply by philippkuehn** [reply]
Hmm maybe you are right. I think we should check for a simple whitespace. Everything else is an edge case we should ignore for now.

**Comment by philippkuehn** [comment]
This sould be fixed with [this path](https://github.com/ueberdosis/tiptap/commit/3f7e6b219b572d2aa06618ce125ba86cd2fcd0bd).
